### DISCUSSION TITLE
I am having an error while i saving trained model as state_dict() ?

### DISCUSSION THREAD

**Discussion by tedzz12** [mrdbourke/pytorch-deep-learning#357] [Q&A] [answered]
this is the code:
from pathlib import Path

MODEL_PATH = Path("models")
MODEL_PATH.mkdir(parents=True, exist_ok=True)

MODEL_NAME = "01_pytorch_workflow_model_1.pth"
MODEL_SAVE_PATH = MODEL_PATH / MODEL_NAME

print(f"Saving model to: {MODEL_SAVE_PATH}")
torch.save(obj=model_1.state_dict(), 
           f=MODEL_SAVE_PATH)

**Comment by mrdbourke** [comment]
Hi @tedzz12,

Let's see what we can do here!

It's very hard to troubleshoot/help without the full code/error messages though.

Also what have you tried so far?

Could you post your code/where you're having an error/what the error is/what you've tried so far?

**Reply by tedzz12** [reply]
```python
from torch.nn.modules.container import ModuleList
from torch.autograd.grad_mode import inference_mode
import torch
from torch import nn
import matplotlib.pyplot as plt

weight = 0.3
bias = 0.9

X = torch.arange(0,3,0.02).unsqueeze(dim=1)
y = weight * X + bias
train_data= int(0.8 * len(X))
X_train,Y_train = X[:train_data],y[:train_data]
X_test,Y_test = X[train_data:],y[train_data:]
len(X_train),len(Y_train),len(X_test),len(Y_test)

def plot_predictions(train_X = X_train,
                     train_Y= Y_train,
                     test_X=X_test,
                     test_Y=Y_test,
                     predictions=None):
  
  plt.figure(figsize=(10,14))

  plt.scatter(train_X,train_Y,c="r",s=3,label="Training Data")
  plt.scatter(test_X,test_Y,c="b",s=3,label="Testing Data")
  if predictions is not None:
    plt.scatter(test_X,predictions,c="y",s=3,label="Predictions")

  plt.legend(prop={"size":14})


class linear_regression(nn.Module):
  def __init__(self):
    super().__init__()
    self.linear_layer = nn.Linear(in_features=1,
                                  out_features=1)
    

  
  def forward(self,x:torch.Tensor)->torch.Tensor:
    return self.linear_layer(x)


model_X = linear_regression()


losfn = nn.L1Loss()
optimizer = torch.optim.SGD(params=model_X.parameters(),
                            lr=0.01)

torch.manual_seed(42)

epochs = 300
epoch_count =[]
train_loss_values =[]
test_loss_values =[]
for epoch in range(epochs):
  model_X.train()
  ypred = model_X(X_train)
  loss = losfn(ypred,Y_train)
  print(f"Loss {loss}")
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()
  model_X.eval()
  with inference_mode():
    epoch_count.append(epoch)
    train_loss_values.append(loss)
    test_loss_values.append(testloss)
    testpred = model_X(X_test)
    testloss = losfn(testpred,Y_test)
  if epoch % 20 == 0:
    print(f"epoch {epoch} | Train Loss {loss} | Test Loss {testloss}")
    print(model_X.state_dict())
  with inference_mode():
    mod_cv = model_X(X_test)
    plot_predictions(predictions=mod_cv)

import numpy as np

np.array(torch.tensor(test_loss_values).cpu().numpy()),train_loss_values

plt.plot(epoch_count,np.array(torch.tensor(test_loss_values).numpy()),label="Test loss values")
plt.plot(epoch_count,np.array(torch.tensor(train_loss_values).numpy()),label="Trained loss values")
plt.title("Train & Test Loss Graph")
plt.xlabel("Epochs")
plt.ylabel("Loss")
plt.legend()

from pathlib import Path


MODEL_PATH = Path("models")
MODEL_PATH.mkdir(parents=True, exist_ok=True)


MODEL_NAME = "01_pytorch_workflow_model_1.pth"
MODEL_SAVE_PATH = MODEL_PATH / MODEL_NAME

print(f"Saving model to: {MODEL_SAVE_PATH}")
torch.save(obj=model_1.state_dict(), 
           f=MODEL_SAVE_PATH)
```

This is the full code, also test loss & loss graphs representing very differently after training 3 or 4 times

**Reply by tedzz12** [reply]
Also I have a doubt, i tried to visualize epoch, test loss and loss  after first training(300 epochs) graph makes sense like mentioned you mentioned in the tutorial, but after 2 or 3 training graph showing very different result.test loss and loss is  going as two parallel lines at top and bottom of the graph.

**Comment by mrdbourke** [comment, marked as answer]
Hi @tedzz12,

Looks like based on your code you're trying to save "`model_1`" where as, your model is called `model_X`:

```python
# Model creation code
model_X = linear_regression()
```

```python
# Save code
print(f"Saving model to: {MODEL_SAVE_PATH}")
torch.save(obj=model_1.state_dict(),  # <- trying to save `model_1` not `model_X`
           f=MODEL_SAVE_PATH)
```

I've updated your code here:

```python
from torch.nn.modules.container import ModuleList
from torch.autograd.grad_mode import inference_mode
import torch
from torch import nn
import matplotlib.pyplot as plt

weight = 0.3
bias = 0.9

X = torch.arange(0,3,0.02).unsqueeze(dim=1)
y = weight * X + bias
train_data= int(0.8 * len(X))
X_train,Y_train = X[:train_data],y[:train_data]
X_test,Y_test = X[train_data:],y[train_data:]
len(X_train),len(Y_train),len(X_test),len(Y_test)

def plot_predictions(train_X = X_train,
                     train_Y= Y_train,
                     test_X=X_test,
                     test_Y=Y_test,
                     predictions=None):
  
  plt.figure(figsize=(10,14))

  plt.scatter(train_X,train_Y,c="r",s=3,label="Training Data")
  plt.scatter(test_X,test_Y,c="b",s=3,label="Testing Data")
  if predictions is not None:
    plt.scatter(test_X,predictions,c="y",s=3,label="Predictions")

  plt.legend(prop={"size":14})


class linear_regression(nn.Module):
  def __init__(self):
    super().__init__()
    self.linear_layer = nn.Linear(in_features=1,
                                  out_features=1)
    

  
  def forward(self,x:torch.Tensor)->torch.Tensor:
    return self.linear_layer(x)


model_X = linear_regression()


losfn = nn.L1Loss()
optimizer = torch.optim.SGD(params=model_X.parameters(),
                            lr=0.01)

torch.manual_seed(42)

epochs = 300
epoch_count =[]
train_loss_values =[]
test_loss_values =[]
for epoch in range(epochs):
  model_X.train()
  ypred = model_X(X_train)
  loss = losfn(ypred,Y_train)
  train_loss_values.append(loss)
  print(f"Loss {loss}")
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()
  model_X.eval()
  with inference_mode():
    epoch_count.append(epoch)
    
    
    testpred = model_X(X_test)
    testloss = losfn(testpred,Y_test)
    test_loss_values.append(testloss)
  if epoch % 20 == 0:
    print(f"epoch {epoch} | Train Loss {loss} | Test Loss {testloss}")
    print(model_X.state_dict())
  with inference_mode():
    mod_cv = model_X(X_test)
    plot_predictions(predictions=mod_cv)

import numpy as np

np.array(torch.tensor(test_loss_values).cpu().numpy()),train_loss_values

plt.plot(epoch_count,np.array(torch.tensor(test_loss_values).numpy()),label="Test loss values")
plt.plot(epoch_count,np.array(torch.tensor(train_loss_values).numpy()),label="Trained loss values")
plt.title("Train & Test Loss Graph")
plt.xlabel("Epochs")
plt.ylabel("Loss")
plt.legend()

from pathlib import Path


MODEL_PATH = Path("models")
MODEL_PATH.mkdir(parents=True, exist_ok=True)


MODEL_NAME = "01_pytorch_workflow_model_1.pth"
MODEL_SAVE_PATH = MODEL_PATH / MODEL_NAME

print(f"Saving model to: {MODEL_SAVE_PATH}")
torch.save(obj=model_X.state_dict(), 
           f=MODEL_SAVE_PATH)
```

Be careful when copying code that the variables you're writing line up.

**Reply by tedzz12** [reply]
Thank you
